src/ProjectSelectionManager.cpp

```diff
@@ -47,91 +47,51 @@ const ProjectSelectionManager &ProjectSelectionManager::Get(
-   : mProject{ project }
-{
-}
-
-ProjectSelectionManager::~ProjectSelectionManager() = default;
-
-bool ProjectSelectionManager::SnapSelection()
-{
-   auto &project = mProject;
-   auto &formats = ProjectNumericFormats::Get( project );
-   auto &settings = ProjectSettings::Get( project );
-   auto &window = ProjectWindow::Get( project );
-   auto snapTo = settings.GetSnapTo();
-   if (snapTo != SNAP_OFF) {
-      auto &viewInfo = ViewInfo::Get( project );
-      auto &selectedRegion = viewInfo.selectedRegion;
-      NumericConverter nc(NumericConverter::TIME,
-         formats.GetSelectionFormat(), 0, ProjectRate::Get(project).GetRate());
-      const bool nearest = (snapTo == SNAP_NEAREST);
-
-      const double oldt0 = selectedRegion.t0();
-      const double oldt1 = selectedRegion.t1();
-
-      nc.ValueToControls(oldt0, nearest);
-      nc.ControlsToValue();
-      const double t0 = nc.GetValue();
-
-      nc.ValueToControls(oldt1, nearest);
-      nc.ControlsToValue();
-      const double t1 = nc.GetValue();
-
-      if (t0 != oldt0 || t1 != oldt1) {
-         selectedRegion.setTimes(t0, t1);
-         return true;
-      }
-   }
-
-   return false;
-}
-
-double ProjectSelectionManager::AS_GetRate()
-{
-   return ProjectRate::Get(mProject).GetRate();
-}
+ProjectSelectionManager::ProjectSelectionManager(AudacityProject& project)
+    : mProject { project }
+    , mSnapModeChangedSubscription { ProjectSnap::Get(project).Subscribe(
+         [this](auto&)
+         {
+            SnapSelection();
+         })}
+    , mTimeSignatureChangedSubscription {
+       ProjectTimeSignature::Get(project).Subscribe([this](auto&)
+                                                    { SnapSelection(); })
+    }
 
-void ProjectSelectionManager::AS_SetRate(double rate)
 {
-   auto &project = mProject;
-   ProjectRate::Get( project ).SetRate( rate );
-   SelectionBar::Get( project ).SetRate(rate);
 }
 
-int ProjectSelectionManager::AS_GetSnapTo()
-{
-   auto &project = mProject;
-   auto &settings = ProjectSettings::Get( project );
-   return settings.GetSnapTo();
-}
+ProjectSelectionManager::~ProjectSelectionManager() = default;
 
-void ProjectSelectionManager::AS_SetSnapTo(int snap)
-{
-   auto &project = mProject;
-   auto &settings = ProjectSettings::Get( project );
-   auto &window = ProjectWindow::Get( project );
+void ProjectSelectionManager::SnapSelection()
+{   
+   auto& project = mProject;
+   auto& projectSnap = ProjectSnap::Get(mProject);
 
-   settings.SetSnapTo( snap );
+   if (projectSnap.GetSnapMode() == SnapMode::SNAP_OFF)
+      return;
 
-// LLL: TODO - what should this be changed to???
-// GetCommandManager()->Check(wxT("Snap"), mSnapTo);
-   gPrefs->Write(wxT("/SnapTo"), snap);
-   gPrefs->Flush();
+   auto& viewInfo = ViewInfo::Get(project);
+   auto& selectedRegion = viewInfo.selectedRegion;
 
-   SnapSelection();
+   const double oldt0 = selectedRegion.t0();
+   const double oldt1 = selectedRegion.t1();
 
-   window.RedrawProject();
```

Comment:
One thing I overlooked in the first round of review.  This `RedrawProject` call was lost.  My branch to break dependencies on toolbars, which conflicts so much with this one, adds a subscription to ProjectWindow so it redraws itself.

Meanwhile: your rebased branch is failing some builds

Reply:
> Meanwhile: your rebased branch is failing some builds

It fails sanity test with the timeout, I will check why.

Reply:
> This RedrawProject call was lost

Is it really need though? I don't know. Fix is quite easy, provided that RedrawProject differs the action.